lib/hls/hls_parser.js

```diff
@@ -2899,59 +2899,59 @@ shaka.hls.HlsParser = class {
-      }
-
-      // Don't download the key object until the segment is parsed, to avoid a
-      // startup delay for long manifests with lots of keys.
-      keyInfo.fetchKey = async () => {
-        const keyUris = shaka.hls.Utils.constructSegmentUris(
-            getUris(), drmTag.getRequiredAttrValue('URI'), variables);
+    // Default AES-128
+    const keyInfo = {
+      bitsKey: 128,
+      blockCipherMode: 'CBC',
+      iv,
+      firstMediaSequenceNumber,
+    };
 
+    const method = drmTag.getRequiredAttrValue('METHOD');
+    switch (method) {
+      case 'AES-256':
+        keyInfo.bitsKey = 256;
+        break;
+      case 'AES-256-CTR':
+        keyInfo.bitsKey = 256;
+        keyInfo.blockCipherMode = 'CTR';
+        break;
+    }
+
+    // Don't download the key object until the segment is parsed, to avoid a
+    // startup delay for long manifests with lots of keys.
+    keyInfo.fetchKey = async () => {
+      const keyUris = shaka.hls.Utils.constructSegmentUris(
+          getUris(), drmTag.getRequiredAttrValue('URI'), variables);
+
+      const keyMapKey = keyUris.sort().join('');
+      if (!this.aesKeyMap_.has(keyMapKey)) {
         const requestType = shaka.net.NetworkingEngine.RequestType.KEY;
         const request = shaka.net.NetworkingEngine.makeRequest(
             keyUris, this.config_.retryParameters);
         const keyResponse =
             await this.makeNetworkRequest_(request, requestType);
+        this.aesKeyMap_.set(keyMapKey, keyResponse.data);
+      }
+      const key = this.aesKeyMap_.get(keyMapKey);
 
-        // keyResponse.status is undefined when URI is "data:text/plain;base64,"
-        if (!keyResponse.data ||
-            keyResponse.data.byteLength != (keyInfo.bitsKey / 8)) {
-          throw new shaka.util.Error(
-              shaka.util.Error.Severity.CRITICAL,
-              shaka.util.Error.Category.MANIFEST,
-              shaka.util.Error.Code.AES_128_INVALID_KEY_LENGTH);
-        }
+      // keyResponse.status is undefined when URI is "data:text/plain;base64,"
+      if (!key || key.byteLength != (keyInfo.bitsKey / 8)) {
+        throw new shaka.util.Error(
+            shaka.util.Error.Severity.CRITICAL,
+            shaka.util.Error.Category.MANIFEST,
+            shaka.util.Error.Code.AES_128_INVALID_KEY_LENGTH);
+      }
 
-        const algorithm = {
-          name: keyInfo.blockCipherMode == 'CTR' ? 'AES-CTR' : 'AES-CBC',
-          length: keyInfo.bitsKey,
-        };
-        keyInfo.cryptoKey = await window.crypto.subtle.importKey(
-            'raw', keyResponse.data, algorithm, true, ['decrypt']);
-        keyInfo.fetchKey = undefined; // No longer needed.
+      const algorithm = {
+        name: keyInfo.blockCipherMode == 'CTR' ? 'AES-CTR' : 'AES-CBC',
+        length: keyInfo.bitsKey,
       };
-      this.aesKeyInfoMap_.set(aesKeyInfoKey, keyInfo);
-    }
-    return this.aesKeyInfoMap_.get(aesKeyInfoKey);
+      keyInfo.cryptoKey = await window.crypto.subtle.importKey(
```

Comment:
Doesn't this mean that, in the case of multiple streams sharing a key, we will call `window.crypto.subtle.importKey` multiple times for the same key?
Perhaps the aesKeyMap_ should contain the result of importKey, instead of the raw buffer of the key.

...I don't know if duplicate calls to importKey cause any problems, admittedly.

Reply:
So it makes sense to keep the previous map and this one at the same time. I will make the change